### PR TITLE
allow recursive data extraction (pydantic program) 

### PR DESCRIPTION
**Pull request by jerryjliu**
# Description

allow for recursive data extraction 

inspired by @jxnl's repo here: https://github.com/jxnl/openai_function_call/blob/main/parse_recursive_paths.py

made fixes to our OpenAIPydanticProgram to make this work

## Type of Change

Please delete options that are not relevant.

- [] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Added new unit/integration tests
- [x] Added new notebook (that tests end-to-end)
- [ ] I stared at the code and made sure it makes sense

### CODE REVIEW

**Review by jxnl — COMMENTED** [review]
It's important to note that by using typed model you lose some type information (which means you lose the ability to autocomplete in IDEs (but not notebooks)

File: docs/examples/output_parsing/directory.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+
+class DirectoryTree(BaseModel):
+    """
+    Container class representing a directory tree.
+
+    Args:
+        root (Node): The root node of the tree.
+
+    Methods:
```
Comment (jxnl) [inline docs/examples/output_parsing/directory.py:45]: No need.

Comment (jxnl) [inline docs/examples/output_parsing/directory.py:45]: Should add the update_forward_ref stuff so class methods can work after extraction.

Comment (jerryjliu) [reply docs/examples/output_parsing/directory.py:45]: done

Comment (jerryjliu) [reply docs/examples/output_parsing/directory.py:45]: done

**Pull request merged**